Nixos19.09

augu5te opened the pull request.

augu5te committed:
[nixos]: minimal from scratch version

augu5te committed:
[nixos] Add nixos-19.09 recipe

augu5te committed:
[nixos-19.09] cleaning configuration.nix

npf commented on steps/bootstrap/prepare_disk.yaml:
```diff
@@ -2,7 +2,7 @@
   - exec_local: |
       rm -f $${image_disk}.$${image_format}
-      qemu-img create -f qcow2 $${image_disk}.$${image_format} $${qemu_image_size}
+      qemu-img create -f $${image_format} $${image_disk}.$${image_format} $${qemu_image_size}
```
Why not using `qcow` but `raw` (as defined in `nixos-*.yaml` recipes) ?
The `image_format` variable is supposed to be the export image format, not the format of the work image. It would be preferable to define a new variable for backward compliance.

augu5te replied:
Need to be tested

npf replied:
reverted

npf commented on from_scratch/nixos-19.03.yaml:
```diff
@@ -0,0 +1,56 @@
+#==============================================================================
+---
+extend: base.yaml
+
```
The wanted scheme is to have a nixos-base.yaml recipe with nixos-19.03 and nixos-19-09 extending it.

augu5te replied:
I don't remenber  why I didn't follow this path.

npf replied:
fixed

npf commented on from_scratch/nixos-19.09.yaml:
```diff
@@ -33,9 +37,20 @@ bootstrap:
+            sudo tar -czf $${appliance_filename}.tgz -C $MNT .
+            sudo chmod 666 $${appliance_filename}.tgz 
+            sudo umount $MNT
+            rm -r $MNT
```
This step makes running the `kameleon build` command require the root privileges, which is not necessary for any other recipe. Rather use the `export_appliance` helper (which uses `guestmount` but does not need root privileges).

augu5te replied:
I was unaware of this

npf replied:
The build VM is not bootable (bootloader installed ?). As a result, guestmount does not work for the export.

Since we would like to possibly export qcow files as well, making the qcow file bootable seems a good idea.

npf replied:
`guestmount` is not looking for a bootloader but for `vmlinuz` and `initrd` files to extract and use directly within `qemu`. This seems to not function with nixos partitioning / non-standard file hierarchy. Need to dive in and find out how to possibly give some insight to `guestmount` with its `-m` option.

npf commented on from_scratch/nixos-19.09.yaml:
```diff
@@ -33,9 +37,20 @@ bootstrap:
   - export:
-    - export_as_qcow2:
-      - exec_local: qemu-img convert -O qcow2 -c $${image_disk}.$${image_format} $${appliance_filename}.qcow2
+    - export_as_tgz:
```
Could the recipe function with the default export method which is defined in the `base.yaml` (just like recipes of other distros), instead of redefining one ? Changing the export could be let to a another recipe extending this recipe (e.g. for Grid'5000).

augu5te replied:
Surely when raw is removed.

npf replied:
removed

npf reviewed (changes requested):
See comments inline.

Todos:
- [ ] rework the recipes extension → `nixos-base.yaml`, then `nixos-<version>.yaml` extending it.
- [ ] fix the qemu disk image format issue
- [ ] fix the export method to use the default one from `base.yaml` or use the correct helper script

npf committed:
[nixos] do not change global script prepare_disk just for nix.

npf committed:
[nixos] rework the recipes inheritance

npf committed:
[nixos] do not change the export section by default

npf commented:
In and Out context are not defined. Only the `qemu-sendkeys` functions for now. The recipe lacks this to enable using the setup section.

npf committed:
[nixos] Move nix data files to the nixos directory

npf committed:
Update the qemy-sendkeys script, so that it handles more characters

npf committed:
[nixos] rework the from_scratch/nixos-base recipe

npf committed:
[nixos] fixup move nixos data files

npf committed:
[base] add option for the first VM boot timeout

npf committed:
Generate the insecure ssh key a bit earlier, before the 1st VM boot

So that the key is available to be sent to the 1st VM possibly.

npf committed:
[nixos] revert to use the base bootstrap.

npf committed:
[nixos] fix the grub installation device

npf committed:
[nixos] make autoinstall.sh install the bootloader

npf committed:
[nixos] fix installer, use new url, add nixos 20.03

npf committed:
Remove g5k workaround for qemu user network

npf committed:
Add option to prepare_appliance to not run the steps using virt-customize

npf committed:
[from_scratch/base] reorganize the bootstrap steps

- Have the ssh keys readly before the installer runs, so that they can
be used right away in the installer.
- Start http before preparing the autoinstall script, so that the
HTTP_PORT is known and can be http service can be used in the
autoinstall script.

npf committed:
[nixos] use kameleon default root password

npf committed:
[nixos] fix the root ssh key and password setup (virt-customize does not work with nixos)

Do not use virt-customize to push the ssh key of root because that does
not work with nixos, push it earlier in the autoinstall script.

npf committed:
[nixos] fix the in_context: bash is /run/current-system/sw/bin/bash

npf committed:
[nixos] do not setup the serial console for grub and linux at this stage.

This should be done before export, not right away, so that debugging in
the qemu vnc console is possible.

npf committed:
[nixos] fix the boot on the root device: qemu needs the virtio driver module
